# Desafío - Manipulación y transformación de datos (Parte I)

## Desarrollo - 1

Para comenzar a trabajar importamos la librería *Pandas*

a. Mediante la creación de cada DataFrame, utilizamos **pd.read_pickle()** para cargar los archivos que contienen los datos para posteriormente visualizar el DataFrame:
- incidents.pkl
- officers.pkl
- subjects.pkl


In [2]:
import pandas as pd

incidents_df = pd.read_pickle('/Users/andrea/Desktop/DATA SCIENCE/Modulo 2-Python/manipulacion y transformacion de datos 1/Apoyo Desafío - Manipulación y transformación de datos (Parte  I)/incidents.pkl')
incidents_df

case_number        date                  location subject_statuses  \
0         44523A  2013-02-23     3000 Chihuahua Street          Injured   
1        121982X  2010-05-03  1300 N. Munger Boulevard          Injured   
2        605484T  2007-08-12   200 S. Stemmons Freeway            Other   
3        384832T  2007-05-26           7900 S. Loop 12   Shoot and Miss   
4        244659R  2006-04-03        6512 South Loop 12          Injured   
..           ...         ...                       ...              ...   
214  161616-2016  2016-07-03   7440 Chesterfield Drive   Shoot and Miss   
215  141461-2016  2016-06-11         2700 Custer Drive   Shoot and Miss   
216  089985-2016  2016-04-16         4800 Columbia Ave   Shoot and Miss   
217  177645-2016  2016-07-23    1716 Chattanooga Place   Shoot and Miss   
218  165193-2016  2016-07-07           801 Main Street         Deceased   

    subject_weapon                                           subjects  \
0          Handgun                                   Curry, James L/M   
1          Handgun                                Chavez, Gabriel L/M   
2          Shotgun                                  Salinas, Nick L/M   
3          Unarmed  Smith, James B/M; Dews, Antonio B/M; Spearman,...   
4            Hands                                 Watkins, Caleb B/M   
..             ...                                                ...   
214        Handgun                                  Brown, Desroy B/M   
215        Vehicle                                        Unknown B/M   
216        Handgun                                        Unknown L/M   
217        Handgun                                        Unknown B/M   
218  Assault Rifle                                 Johnson, Micah B/M   

     subject_count                                           officers  \
0                1          Patino, Michael L/M; Fillingim, Brian W/M   
1                1                               Padilla, Gilbert L/M   
2                1                                  Poston, Jerry W/M   
3                3                                 Mondy, Michael B/M   
4                1                             Armstrong, Michael W/M   
..             ...                                                ...   
214              1                               Crenshaw, Joshua W/M   
215              1                                 Moore, Douglas B/M   
216              1                               Ruben, Fredirick W/M   
217              1                                   Rosen, Brett W/M   
218              1  Edwards, Henry W/M; Wells, Giovanni B/M; Junge...   

     officer_count grand_jury_disposition attorney_general_forms_url  \
0                2                No Bill                       None   
1                1                No Bill                       None   
2                1            See Summary                       None   
3                1                   None                       None   
4                1                No Bill                       None   
..             ...                    ...                        ...   
214              1                   None                       None   
215              1                   None                       None   
216              1                   None                       None   
217              1                   None                       None   
218             12                   None                       None   

                                           summary_url  \
0    http://dallaspolice.net/reports/OIS/narrative/...   
1    http://dallaspolice.net/reports/OIS/narrative/...   
2    http://dallaspolice.net/reports/OIS/narrative/...   
3    http://dallaspolice.net/reports/OIS/narrative/...   
4    http://dallaspolice.net/reports/OIS/narrative/...   
..                                                 ...   
214  http://dallaspolice.net/reports/OIS/narrative/...   
215  http://dallaspol

In [2]:
officers_df = pd.read_pickle('/Users/andrea/Desktop/DATA SCIENCE/Modulo 2-Python/manipulacion y transformacion de datos 1/Apoyo Desafío - Manipulación y transformación de datos (Parte  I)/officers.pkl')
officers_df

case_number race gender  last_name  first_name           full_name
0         44523A    L      M     Patino     Michael     Patino, Michael
1         44523A    W      M  Fillingim       Brian    Fillingim, Brian
2        121982X    L      M    Padilla     Gilbert    Padilla, Gilbert
3        605484T    W      M     Poston       Jerry       Poston, Jerry
4        384832T    B      M      Mondy     Michael      Mondy, Michael
..           ...  ...    ...        ...         ...                 ...
365  165193-2016    W      M   Michaels        Mark      Michaels, Mark
366  165193-2016    W      M  Borchardt      Jeremy   Borchardt, Jeremy
367  165193-2016    W      M      Craig      Robert       Craig, Robert
368  165193-2016    W      M     Cannon       Elmar       Cannon, Elmar
369  165193-2016    B      M     Stuart  Kristopher  Stuart, Kristopher

[370 rows x 6 columns]

In [3]:
subjects_df = pd.read_pickle('/Users/andrea/Desktop/DATA SCIENCE/Modulo 2-Python/manipulacion y transformacion de datos 1/Apoyo Desafío - Manipulación y transformación de datos (Parte  I)/subjects.pkl')
subjects_df

case_number race gender last_name first_name        full_name
0         44523A    L      M     Curry      James     Curry, James
1        121982X    L      M    Chavez    Gabriel  Chavez, Gabriel
2        605484T    L      M   Salinas       Nick    Salinas, Nick
3        384832T    B      M     Smith      James     Smith, James
4        384832T    B      M      Dews    Antonio    Dews, Antonio
..           ...  ...    ...       ...        ...              ...
218  161616-2016    B      M     Brown     Desroy    Brown, Desroy
219  141461-2016    B      M   Unknown       None          Unknown
220  089985-2016    L      M   Unknown       None          Unknown
221  177645-2016    B      M   Unknown       None          Unknown
222  165193-2016    B      M   Johnson      Micah   Johnson, Micah

[223 rows x 6 columns]

b. Generamos una tabla que contenga la unión de las 3 tablas utilizando sufijos para las columnas con nombres iguales. Para ello debemos unir primero dos tablas y después unir el resulado con la tercera tabla:

- Primero se unirá la tabla *subjects* con *incidents* formando la tabla *subjects_incidents_df*. 

- Segundo de unirá la tabla *subjects_incidents_df* con *officers*, para formar finalmente la tabla **merged_df**.

La unión de las tablas se hacen mediante la columna **case_number** que es la columna que tienen en común las 3 tablas.

En el parámetro **how** usamos el valor **outer** porque necesitamos todos los datos contenidos en las 3 tablas.

-> Hacemos el uso del prámetro **suffixes** que se usa cuando se fusionan dos DataFrames que tienen columnas con el mismo nombre. Su uso es diferenciar estas columnas en el DataFrame resultante. Este parámetro toma una tupla de dos strings que serán añadidos como sufijos a las columnas de cada DataFrame. La tupla *('_df1', '_df2')* especifica los sufijos siendo el primer string '_df1' aplicado a las columnas duplicadas de df1 (por ejemplo subjects_df), y el segundo string '_df2' aplicado a las columnas duplicadas de df2 (por ejemplo incidents_df).

In [4]:
# Unir subjects con incidents

subjects_incidents_df = pd.merge(subjects_df, incidents_df, on='case_number', how='outer', suffixes=('_subject', '_incident'))

# Unir el resultado anterior con officers

merged_df = pd.merge(subjects_incidents_df, officers_df, on='case_number',  how='outer', suffixes=('_subject', '_officer'))

# Mostrar la tabla unida

merged_df

case_number race_subject gender_subject last_name_subject  \
0         44523A            L              M             Curry   
1         44523A            L              M             Curry   
2        121982X            L              M            Chavez   
3        605484T            L              M           Salinas   
4        384832T            B              M             Smith   
..           ...          ...            ...               ...   
375  165193-2016            B              M           Johnson   
376  165193-2016            B              M           Johnson   
377      203505V          NaN            NaN               NaN   
378      199351X          NaN            NaN               NaN   
379       62683Z          NaN            NaN               NaN   

    first_name_subject full_name_subject        date  \
0                James      Curry, James  2013-02-23   
1                James      Curry, James  2013-02-23   
2              Gabriel   Chavez, Gabriel  2010-05-03   
3                 Nick     Salinas, Nick  2007-08-12   
4                James      Smith, James  2007-05-26   
..                 ...               ...         ...   
375              Micah    Johnson, Micah  2016-07-07   
376              Micah    Johnson, Micah  2016-07-07   
377                NaN               NaN  2008-06-27   
378                NaN               NaN  2010-07-16   
379                NaN               NaN  2012-03-15   

                     location subject_statuses subject_weapon  ...  \
0       3000 Chihuahua Street          Injured        Handgun  ...   
1       3000 Chihuahua Street          Injured        Handgun  ...   
2    1300 N. Munger Boulevard          Injured        Handgun  ...   
3     200 S. Stemmons Freeway            Other        Shotgun  ...   
4             7900 S. Loop 12   Shoot and Miss        Unarmed  ...   
..                        ...              ...            ...  ...   
375           801 Main Street         Deceased  Assault Rifle  ...   
376           801 Main Street         Deceased  Assault Rifle  ...   
377       2900 Wheatland Road            Other        Handgun  ...   
378         3701 Crane Street          Injured        Handgun  ...   
379   400 S. Llewellyn Avenue          Injured        Unarmed  ...   

    attorney_general_forms_url  \
0                         None   
1                         None   
2                         None   
3                         None   
4                         None   
..                         ...   
375                       None   
376                       None   
377                       None   
378                       None   
379                       None   

                                           summary_url  \
0    http://dallaspolice.net/reports/OIS/narrative/...   
1    http://dallaspolice.net/reports/OIS/narrative/...   
2    http://dallaspolice.net/reports/OIS/narrative/...   
3    http://dallaspolice.net/reports/OIS/narrative/...   
4    http://dallaspolice.net/reports/OIS/narrative/...   
..                                                 ...   
375                                               None   
376                                               None   
377  http://dallaspolice.net/reports/OIS/narrative/...   
378  http://dallaspolice.net/reports/OIS/narrative/...   
379  http://dallaspolice.net/reports/OIS/narrative/...   

                                          summary_text  latitude  longitude  \
0    On Saturday, February 23, 2013, at approximate...       NaN        NaN   
1    On Saturday, February 23, 2013, at approximate...       NaN        NaN   
2    On Monday, May 3, 2010, at approximately 8:06 ...       NaN        NaN   
3    On Sunday, August 12, 2007, at approximately 2...       NaN        NaN   
4    On Saturday, May 26, 2007, at approximately 1:...       NaN        NaN   
..                                                 ...       ...        ...   
375                                    

c. Verificamos si hay filas duplicadas y las eliminamos

- Primero verificamos si hay tablas duplicadas con **.duplicated()** y mostramos el resultado mediante *print()* y con *sum()* contamos el total de los datos duplicados.

- Según el paso anterior no hay datos duplicados, pero en caso de que se modifiquen los datos a futuro, los datos duplicados se eliminan con **.drop_duplicates()**. Con *len()* podemos demostar la cantidad de datos que quedaron en *merged_df* después de eliminar los duplicados. 

In [5]:
# Verificar duplicados

filas_duplicadas = merged_df.duplicated()
print(f'Filas duplicadas: {filas_duplicadas.sum()}')

# Eliminar duplicados

merged_df = merged_df.drop_duplicates()
print(f'Filas después de eliminar duplicados: {len(merged_df)}')

Filas duplicadas: 0
Filas después de eliminar duplicados: 380


d. Contamos cuantos sujetos de genero femenino tenemos, para ello usamos **value_counts()** para contar las ocurrencias de cada valor único en la columna *gender_subject*, por otro lado con **.get('F', 0)** obtenemos el conteo para el valor específico 'F'. Si 'F' no está presente en la Serie, get devuelve el valor predeterminado 0 en lugar de lanzar un error.

In [6]:
# Contar sujetos de género F

female_count = merged_df['gender_subject'].value_counts().get('F', 0)
print(f'Sujetos de género F: {female_count}')

Sujetos de género F: 9


e. Contamos números de caso con al menos una sospechosa mujer. Usamos:

- **query()** para filtrar las filas del DataFrame merged_df donde la columna *gender_subject* es igual a 'F'.

- **['case_number']** para seleccionar la columna *case_number* de las filas filtradas.

- **.unique()** para obtener los valores únicos de la columna case_number filtrada.


In [7]:
# Filtrar casos con al menos una mujer sospechosa

female_suspect = merged_df[merged_df['gender_subject'] == 'F']['case_number'].unique()
print(f'Casos con al menos una sospechosa mujer: {len(female_suspect)}')

Casos con al menos una sospechosa mujer: 7


f.Generamos una tabla pivote para género del oficial y del sujeto. Usamos:

- **pd.pivot_table**: Llama a la función pivot_table para crear una tabla dinámica a partir del DataFrame *merged_df*.

- **index**: Especifica la columna que se utilizará como índice tabla. En este caso, *gender_officer* que representa el género de los oficiales.

- **columns**: Especifica la columna que se utilizará como columnas en la tabla, acá es *gender_subject*, que representa el género de los sujetos.

- **values**: Especifica la columna cuyos valores se utilizarán en la tabla. En este caso, *case_number* que representa los números de caso.

- **aggfunc**: Especifica la función de agregación que se utilizará para calcular los valores, usaremos **'count'** para contar el número de ocurrencias de *case_number* para cada combinación de *gender_officer* y *gender_subject*.


In [8]:
# Crear tabla pivote

pivot_table = pd.pivot_table(merged_df 
                             ,index='gender_officer'
                             ,columns='gender_subject'
                             ,values='case_number'
                             ,aggfunc='count')

pivot_table

gender_subject  F    M
gender_officer        
F               2   18
M               7  350

### Interpretación de los datos

#### Interpretación de la Tabla Pivote

La tabla pivote muestra el número de casos agrupados por el género de los oficiales (gender_officer) y el género de los sujetos (gender_subject). Los valores son los siguientes:

- Oficiales Femeninos (F) y Sujetos Femeninos (F): 2 casos
- Oficiales Femeninos (F) y Sujetos Masculinos (M): 18 casos
- Oficiales Masculinos (M) y Sujetos Femeninos (F): 7 casos
- Oficiales Masculinos (M) y Sujetos Masculinos (M): 350 casos

#### Análisis:

- Predominio de Oficiales Masculinos: Los oficiales masculinos tienen un número significativamente mayor de casos en comparación con las oficiales femeninas.

- Interacciones de Género: La mayoría de los casos involucran oficiales masculinos y sujetos masculinos con un total de 350 casos. Los oficiales masculinos interactúan con sujetos femeninos en 7 casos. Por otro lado, las oficiales femeninas están involucradas en 20 casos en total, dividiéndose de la siguiente manera: 2 con sujetos femeninos y 18 con sujetos masculinos.

- Oficiales Femeninos y Sujetos Femeninos: El número de casos que involucran tanto a oficiales como a sujetos femeninos es muy bajo, con 2 casos en total, lo que puede indicar dos cosas: menos frecuencia de estas interacciones o una representación desproporcionadamente baja de oficiales femeninas.

Estos datos pueden ser útiles para entender las dinámicas de género dentro de los incidentes reportados y pueden proporcionar información valiosa para iniciativas de diversidad y capacitación dentro de la policia.


## Desarrollo - 2

a. Para comenzar con el desarrollo del segundo ejercicio primero importamos la librería NumPy ya que para el desarrollo de éste necesitaremos usar funciones de esta librería en conjunto con la ya importada de Pandas. En segundo lugar, cargamos los datos y creamos el DataFrame *trabajos_df*

In [3]:
import numpy as np

trabajos_df = pd.read_csv('/Users/andrea/Desktop/DATA SCIENCE/Modulo 2-Python/manipulacion y transformacion de datos 1/Apoyo Desafío - Manipulación y transformación de datos (Parte  I)/Cleaned_DS_Jobs.csv')

trabajos_df.head(10)

Job Title Salary Estimate  \
0                         Sr Data Scientist        137-171    
1                            Data Scientist        137-171    
2                            Data Scientist        137-171    
3                            Data Scientist        137-171    
4                            Data Scientist        137-171    
5                            Data Scientist        137-171    
6  Data Scientist / Machine Learning Expert        137-171    
7                            Data Scientist        137-171    
8          Staff Data Scientist - Analytics        137-171    
9                            Data Scientist        137-171    

                                     Job Description  Rating  \
0  Description\n\nThe Senior Data Scientist is re...     3.1   
1  Secure our Nation, Ignite your Future\n\nJoin ...     4.2   
2  Overview\n\n\nAnalysis Group is one of the lar...     3.8   
3  JOB DESCRIPTION:\n\nDo you have a passion for ...     3.5   
4  Data Scientist\nAffinity Solutions / Marketing...     2.9   
5  About Us:\n\nHeadquartered in beautiful Santa ...     4.2   
6  Posting Title\nData Scientist / Machine Learni...     3.9   
7  Introduction\n\nHave you always wanted to run ...     3.5   
8  Intuit is seeking a Staff Data Scientist to co...     4.4   
9  Ready to write the best chapter of your career...     3.6   

         Company Name           Location            Headquarters  \
0         Healthfirst       New York, NY            New York, NY   
1             ManTech      Chantilly, VA             Herndon, VA   
2      Analysis Group         Boston, MA              Boston, MA   
3             INFICON         Newton, MA  Bad Ragaz, Switzerland   
4  Affinity Solutions       New York, NY            New York, NY   
5         HG Insights  Santa Barbara, CA       Santa Barbara, CA   
6            Novartis      Cambridge, MA      Basel, Switzerland   
7              iRobot        Bedford, MA             Bedford, MA   
8       Intuit - Data      San Diego, CA       Mountain View, CA   
9  XSELL Technologies        Chicago, IL             Chicago, IL   

                      Size        Type of ownership  \
0   1001 to 5000 employees   Nonprofit Organization   
1  5001 to 10000 employees         Company - Public   
2   1001 to 5000 employees  Private Practice / Firm   
3    501 to 1000 employees         Company - Public   
4      51 to 200 employees        Company - Private   
5      51 to 200 employees        Company - Private   
6         10000+ employees         Company - Public   
7   1001 to 5000 employees         Company - Public   
8  5001 to 10000 employees         Company - Public   
9      51 to 200 employees        Company - Private   

                                   Industry  ... company_age python  excel  \
0                        Insurance Carriers  ...          27      0      0   
1                    Research & Development  ...          52      0      0   
2                                Consulting  ...          39      1      1   
3     Electrical & Electronic Manufacturing  ...          20      1      1   
4                   Advertising & Marketing  ...          22      1      1   
5              Computer Hardware & Software  ...          10      1      1   
6                 Biotech & Pharmaceuticals  ...          24      1      0   
7  Consumer Electronics & Appliances Stores  ...          30      1      0   
8              Computer Hardware & Software  ...          37      0      0   
9   Enterprise Software & Network Solutions  ...           6      1      0   

   hadoop  spark aws  tableau  big_data        job_simp  seniority  
0       0      0   1        0         0  data scientist     senior  
1       1      0   0        0         1  data scientist         na  
2       0      0   1        0         0  data scientist         na  
3       0      0   1        0         0  data scientist         na  
4       0      0   0        0         0  data scientist         na  
5      

b. Reemplazamos los valoes nulos. Para ello:

- Usamos la lista **valores_nulos** que crea una lista de valores que consideramos como nulos.

- **replace(valores_nulos, np.nan)**: Utilizamos el método **replace** para reemplazar todos los valores en la lista *valores_nulos* por **np.nan** que representa valores nulos en NumPy.

- **inplace=True**: Indica que los cambios se deben hacer directamente en el DataFrame trabajos_df.

- **isna()**: Es un método para crear un DataFrame booleano del mismo tamaño que *trabajos_df*, donde 'True' indica la presencia de un valor nulo (np.nan) y 'False' indica la presencia de un valor no nulo.

- **.sum()**: Suma los valores nulos a lo largo de cada columna.


In [11]:
# Lista de valores considerados como nulos
valores_nulos = ["na", "NA", -1, "0", "-1", "null", "n/a", "N/A", "NULL"]

# Reemplazar valores nulos
trabajos_df.replace(valores_nulos, np.nan, inplace=True)

# Verificar valores nulos
print(trabajos_df.isna().sum())

Job Title              0
Salary Estimate        0
Job Description        0
Rating                 0
Company Name           0
Location               0
Headquarters          31
Size                  27
Type of ownership     27
Industry              71
Sector                71
Revenue               27
min_salary             0
max_salary             0
avg_salary             0
job_state              0
same_state             0
company_age          118
python                 0
excel                  0
hadoop                 0
spark                  0
aws                    0
tableau                0
big_data               0
job_simp              68
seniority            567
dtype: int64


c. Eliminar filas con datos faltantes. Para ello usamos:

- **dropna()**: Este método se utiliza para eliminar todas las filas que contienen al menos un valor nulo (NaN).

- **reset_index()**: Este método se utiliza para restablecer el índice del DataFrame. En este caso lo usamos para reindexar después de haber eliminado los valores nulos.

- **drop=True**: Indica que el índice anterior debe ser eliminado en lugar de ser agregado como una columna en el DataFrame resultante.


In [4]:
# Eliminar filas con datos faltantes
trabajos_df.dropna().reset_index(drop= True, inplace=True)

# Verificar el DataFrame después de eliminar filas con datos faltantes
trabajos_df.isna().sum()

Job Title            0
Salary Estimate      0
Job Description      0
Rating               0
Company Name         0
Location             0
Headquarters         0
Size                 0
Type of ownership    0
Industry             0
Sector               0
Revenue              0
min_salary           0
max_salary           0
avg_salary           0
job_state            0
same_state           0
company_age          0
python               0
excel                0
hadoop               0
spark                0
aws                  0
tableau              0
big_data             0
job_simp             0
seniority            0
dtype: int64

d. Generar columnas de salario mínimo y máximo estimados. Para ello:

- Usamos **apply()** para aplicar una función a cada elemento de la columna *Salary Estimate*.

- **lambda** define una función que toma un argumento **x**.

- **x.split("-")**: *split("-")* divide la cadena **x** en una lista de dos elementos, separando por el carácter "-"-". 

- **[float(i.strip()) for i in x.split("-")]**: Esta es la función que va a aplicar *lambda*, ésta recorre cada elemento **i** en la lista de *x.split("-")*, además quita espacios en blanco alrededor de **i** con **strip()**, y convierte **i** a un float.

- **[0]** selecciona el primer elemento de la lista resultante, que es el salario mínimo.

- Para **Salario Máximo Estimado** usamos el mismo código pero en vez de utilizar 0 utilizamos **[1]**, ya que selecciona el segundo elemento de la lista, que corresponde al salario máximo.



In [5]:
# Crear la columna Salario Mínimo Estimado
trabajos_df["Salario Mínimo Estimado"] = trabajos_df["Salary Estimate"].apply(lambda x: [float(i.strip()) for i in x.split("-")][0])

#Crear la columna Salario Máximo Estimado
trabajos_df["Salario Máximo Estimado"] = trabajos_df["Salary Estimate"].apply(lambda x: [float(i.strip()) for i in x.split("-")][1])

#Mostramos el DataFrame para verificar los cambios realizados hasta ahora
trabajos_df

Job Title Salary Estimate  \
0    Sr Data Scientist        137-171    
1       Data Scientist        137-171    
2       Data Scientist        137-171    
3       Data Scientist        137-171    
4       Data Scientist        137-171    
..                 ...             ...   
655     Data Scientist        105-167    
656     Data Scientist        105-167    
657     Data Scientist        105-167    
658     Data Scientist        105-167    
659     Data Scientist        105-167    

                                       Job Description  Rating  \
0    Description\n\nThe Senior Data Scientist is re...     3.1   
1    Secure our Nation, Ignite your Future\n\nJoin ...     4.2   
2    Overview\n\n\nAnalysis Group is one of the lar...     3.8   
3    JOB DESCRIPTION:\n\nDo you have a passion for ...     3.5   
4    Data Scientist\nAffinity Solutions / Marketing...     2.9   
..                                                 ...     ...   
655  Summary\n\nWe’re looking for a data scientist ...     3.6   
656  Job Description\nBecome a thought leader withi...     0.0   
657  Join a thriving company that is changing the w...     0.0   
658  100 Remote Opportunity As an AINLP Data Scient...     5.0   
659  Description\n\nThe Data Scientist will be part...     2.7   

             Company Name           Location            Headquarters  \
0             Healthfirst       New York, NY            New York, NY   
1                 ManTech      Chantilly, VA             Herndon, VA   
2          Analysis Group         Boston, MA              Boston, MA   
3                 INFICON         Newton, MA  Bad Ragaz, Switzerland   
4      Affinity Solutions       New York, NY            New York, NY   
..                    ...                ...                     ...   
655              TRANZACT       Fort Lee, NJ            Fort Lee, NJ   
656                  JKGT  San Francisco, CA                      -1   
657            AccessHope      Irwindale, CA                      -1   
658  ChaTeck Incorporated  San Francisco, CA         Santa Clara, CA   
659         1-800-Flowers       New York, NY         Carle Place, NY   

                        Size        Type of ownership  \
0     1001 to 5000 employees   Nonprofit Organization   
1    5001 to 10000 employees         Company - Public   
2     1001 to 5000 employees  Private Practice / Firm   
3      501 to 1000 employees         Company - Public   
4        51 to 200 employees        Company - Private   
..                       ...                      ...   
655   1001 to 5000 employees        Company - Private   
656                       -1                       -1   
657                       -1                       -1   
658        1 to 50 employees        Company - Private   
659   1001 to 5000 employees         Company - Public   

                                  Industry  ... excel hadoop  spark  aws  \
0                       Insurance Carriers  ...     0      0      0    1   
1                   Research & Development  ...     0      1      0    0   
2                               Consulting  ...     1      0      0    1   
3    Electrical & Electronic Manufacturing  ...     1      0      0    1   
4                  Advertising & Marketing  ...     1      0      0    0   
..                                     ...  ...   ...    ...    ...  ...   
655                Advertising & Marketing  ...     1      1      0    0   
656                                     -1  ...     0      0      0    0   
657                                     -1  ...     1      1      0    0   
658                Advertising & Marketing  ...     0      1      1    0   
659                              Wholesale  ...     1      0      0    0   

     tableau big_data        job_simp  seniority  Salario Mínimo Estimado  \
0          0        0  data scientist     senior                    137.0   
1          0        1  data scientist         na                    137.0   
2          0        0

e. Recodificación de la columna Size. 
  
Para poder cambiar los valores de la columna 'Size' creamos un diccionario llamado *size_recod*.**trabajos_df['Size']** que se usa para seleccionar la columna "Size".Por otro lado, **replace(size_recod, inplace=True)** usa el diccionario para reemplazar los valores en la columna "Size" con los nuevos valores. 


In [6]:
#Crear diccionario
size_recod = {
    "10000+ employees": "Mega Empresas",
    "5001 to 10000 employees": "Grandes Empresas",
    "1001 to 5000 employees": "Medianas Empresas",
    "201 to 500 employees": "Pequeñas Empresas",
    "51 to 200 employees": "Pequeñas Grandes Empresas",
    "501 to 1000 employees": "Microempresas",
    "Unknown": "Empresas sin Información"
}

#Reemplazar los valores
trabajos_df['Size'].replace(size_recod, inplace=True)

#Verificar el cambio 
trabajos_df


Job Title Salary Estimate  \
0    Sr Data Scientist        137-171    
1       Data Scientist        137-171    
2       Data Scientist        137-171    
3       Data Scientist        137-171    
4       Data Scientist        137-171    
..                 ...             ...   
655     Data Scientist        105-167    
656     Data Scientist        105-167    
657     Data Scientist        105-167    
658     Data Scientist        105-167    
659     Data Scientist        105-167    

                                       Job Description  Rating  \
0    Description\n\nThe Senior Data Scientist is re...     3.1   
1    Secure our Nation, Ignite your Future\n\nJoin ...     4.2   
2    Overview\n\n\nAnalysis Group is one of the lar...     3.8   
3    JOB DESCRIPTION:\n\nDo you have a passion for ...     3.5   
4    Data Scientist\nAffinity Solutions / Marketing...     2.9   
..                                                 ...     ...   
655  Summary\n\nWe’re looking for a data scientist ...     3.6   
656  Job Description\nBecome a thought leader withi...     0.0   
657  Join a thriving company that is changing the w...     0.0   
658  100 Remote Opportunity As an AINLP Data Scient...     5.0   
659  Description\n\nThe Data Scientist will be part...     2.7   

             Company Name           Location            Headquarters  \
0             Healthfirst       New York, NY            New York, NY   
1                 ManTech      Chantilly, VA             Herndon, VA   
2          Analysis Group         Boston, MA              Boston, MA   
3                 INFICON         Newton, MA  Bad Ragaz, Switzerland   
4      Affinity Solutions       New York, NY            New York, NY   
..                    ...                ...                     ...   
655              TRANZACT       Fort Lee, NJ            Fort Lee, NJ   
656                  JKGT  San Francisco, CA                      -1   
657            AccessHope      Irwindale, CA                      -1   
658  ChaTeck Incorporated  San Francisco, CA         Santa Clara, CA   
659         1-800-Flowers       New York, NY         Carle Place, NY   

                          Size        Type of ownership  \
0            Medianas Empresas   Nonprofit Organization   
1             Grandes Empresas         Company - Public   
2            Medianas Empresas  Private Practice / Firm   
3                Microempresas         Company - Public   
4    Pequeñas Grandes Empresas        Company - Private   
..                         ...                      ...   
655          Medianas Empresas        Company - Private   
656                         -1                       -1   
657                         -1                       -1   
658          1 to 50 employees        Company - Private   
659          Medianas Empresas         Company - Public   

                                  Industry  ... excel hadoop  spark  aws  \
0                       Insurance Carriers  ...     0      0      0    1   
1                   Research & Development  ...     0      1      0    0   
2                               Consulting  ...     1      0      0    1   
3    Electrical & Electronic Manufacturing  ...     1      0      0    1   
4                  Advertising & Marketing  ...     1      0      0    0   
..                                     ...  ...   ...    ...    ...  ...   
655                Advertising & Marketing  ...     1      1      0    0   
656                                     -1  ...     0      0      0    0   
657                                     -1  ...     1      1      0    0   
658                Advertising & Marketing  ...     0      1      1    0   
659                              Wholesale  ...     1      0      0    0   

     tableau big_data        job_simp  seniority  Salario Mínimo Estimado  \
0          0        0  data scientist     senior                    137.0   
1          0        1  data scientist         na                    137.0 

f. Generar tabla pivote para el salario por tamaño de empresa. Explicaré paso a paso:

- **Asignación del resultado**: Lo realizamos mediante la variable **salario_pivote = ...**

- **trabajos_df.pivot_table**: Usamos esta función que se utiliza para crear una tabla pivote.

- **index**: Este parámetro especifica la columna que se utilizará como índice de la tabla pivote. En este caso, estamos utilizando la columna **Size** que representa el tamaño de la empresa.

- **values**:Este parámetro especifica las columnas cuyos valores serán agregados en la tabla pivote. Aquí, estamos usando las columnas Salario Mínimo Estimado y Salario Máximo Estimado.

- **aggfunc**: Este parámetro especifica la función de agregación que se usará para resumir los datos. En este caso usamos **aggfunc='mean'**, el cual indica que se calculará la media de los valores de *Salario Mínimo Estimado* y *Salario Máximo Estimado* para cada grupo de Size.


In [ ]:
# d. A partir de la columna “Salary Estimate”, genera dos columnas: Salario Estimado Mínimo y Máximo.
# (hint: Utiliza el método apply sobre la columna.)

# Dividir la columna "Salary Estimate" en mínimo y máximo
salarios_min_max = cleaned_ds_jobs_df['Salary Estimate'].str.split('-', expand=True)

# Asignar las nuevas columnas al DataFrame original
cleaned_ds_jobs_df['Salario Estimado Mínimo'] = salarios_min_max[0].str.extract('(\d+)', expand=False).astype(float) * 1000
cleaned_ds_jobs_df['Salario Estimado Máximo'] = salarios_min_max[1].str.extract('(\d+)', expand=False).astype(float) * 1000


cleaned_ds_jobs_df